### PR TITLE
[SYSTEMML-1872] Added average pooling and upsampling layers

### PR DESCRIPTION
**Pull request by niketanpansare**
- Added avg_pool and avg_pool_backward builtin functions.
- The above builtin functions are supported on both CPU and GPU.
- Also, added compare_backend tests for the above function to compare the results of CP operator with that of CuDNN.
- Added avg_pool2d_builtin.dml and upsample2d.dml in the nn library.
- Added gradcheck tests for the above layers.
- Supported average pooling and upsampling in Keras2DML as well as Caffe2DML.
- Tested the results of ResNet with average pooling on real-world images as a sanity check.
- Also, tested upsampling layer by comparing the results with that returned by Keras.

@prithvirajsen @bertholdreinwald @dusenberrymw

### CODE REVIEW

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2288/

File: scripts/nn/layers/upsample2d.dml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+   *  - size_w: upsampling factor for columns.
+   *
+   * Outputs:
+   *  - out: Outputs, of shape [size_h*Hin, size_w*Win].
+   */
+  N = nrow(X)
+  Hout = size_h*Hin
+  Wout = size_w*Win
+  emptyInput = matrix(0, rows=N, cols=C*Hout*Wout)
+  out = avg_pool_backward(emptyInput, X, input_shape=[N,C,Hout,Wout], pool_size=[size_h,size_w], stride=[size_h,size_w], padding=[0,0])
```
Comment (dusenberrymw) [inline scripts/nn/layers/upsample2d.dml]: This is interesting -- math-wise the avg pool gradient doesn't need the input values anyway so long as the shape is known/computed.  I suppose we could consider dropping that parameter from the built-in op, although the API consistency is nice (we'd want to keep it in the dml layer regardless for consistency) -- does cuDNN use it?

Comment (niketanpansare) [reply scripts/nn/layers/upsample2d.dml]: Yes, both cudnn maxpool and avg pool have same signature

Comment (dusenberrymw) [reply scripts/nn/layers/upsample2d.dml]: Alright, might as well keep it for consistency then!

File: scripts/nn/layers/upsample2d.dml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+forward = function(matrix[double] X, int C, int Hin, int Win, int size_h, int size_w)
+    return (matrix[double] out) {
+  /*
+   * Computes the forward pass for a Upsampling layer.
+   *
+   *
+   * Inputs:
+   *  - X: Inputs, of shape (any, any).
```
[the remark this answers is not in the document]
Comment (niketanpansare) [reply scripts/nn/layers/upsample2d.dml]: Done.

File: scripts/nn/layers/upsample2d.dml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+   *  - C: Number of input channels (dimensionality of input depth).
+   *  - Hin: Input height.
+   *  - Win: Input width.
+   *  - size_h: upsampling factor for rows.
+   *  - size_w: upsampling factor for columns.
+   *
+   * Outputs:
+   *  - out: Outputs, of shape [size_h*Hin, size_w*Win].
```
[the remark this answers is not in the document]
Comment (niketanpansare) [reply scripts/nn/layers/upsample2d.dml]: Done.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2290/

**Pull request closed without merging**